File: java/java-basic/src/main/java/me/hvkcoder/java_basic/data_struct/_LinkedList.java
package me.hvkcoder.java_basic.data_struct;

import me.hvkcoder.java_basic.data_struct.base._AbstractList;

/**
 * @author h-vk
 * @since 2020/11/30
 */
public class _LinkedList<E> extends _AbstractList<E> {

	/**
	 * 链表头
	 */
	private Node<E> head;

	/**
	 * 链表尾
	 */
	private Node<E> last;

	/**
	 * 链表 Node 结构
	 *
	 * @param <E>
	 */
	private class Node<E> {
		private E data;
		private Node<E> prev;
		private Node<E> next;

		public Node(E data) {
			this.data = data;
		}

		public Node(E data, Node<E> prev, Node<E> next) {
			this.data = data;
			this.prev = prev;
			this.next = next;
		}

		@Override
		public String toString() {
			return "Node{" + "data=" + data + ", next=" + next + '}';
		}
	}

	@Override
	public void clear() {
		head = null;
		last = null;
		size = 0;
	}

	@Override
	public void add(int index, E element) {
		checkedIndexForAdd(index);
		Node<E> newNode = new Node<>(element);
		if (size == 0) {
			head = newNode;
			last = newNode;
		} else if (index == 0) {
			newNode.next = head;
			head = newNode;
		} else if (index == size) {
			last.next = newNode;
			last = newNode;
		} else {
			Node<E> prevNode = getNode(index - 1);
			newNode.next = prevNode.next;
			prevNode.next = newNode;
		}
		size++;
	}

	/**
	 * 根据 index 获取 Node 节点值
	 *
	 * @param index
	 * @return
	 */
	@Override
	public E get(int index) {
		return getNode(index).data;
	}

	/**
	 * 根据 index 获取 Node 节点
	 *
	 * @param index
	 * @return
	 */
	public Node<E> getNode(int index) {
		checkedIndex(index);
		Node<E> node = head;
		for (int i = 0; i < index; i++) {
			node = node.next;
		}
		return node;
	}

	@Override
	public E set(int index, E element) {
		Node<E> node = getNode(index);
		E oldValue = node.data;
		node.data = element;
		return oldValue;
	}

	@Override
	public E remove(int index) {
		checkedIndex(index);
		Node<E> removeNode = head;
		if (index == 0) {
			head = removeNode.next;
		} else if (index == size - 1) {
			Node<E> prevNode = getNode(index - 1);
			removeNode = last;
			prevNode.next = null;
			last = prevNode;
		} else {
			Node<E> prevNode = getNode(index - 1);
			removeNode = prevNode.next;
			prevNode.next = removeNode.next;
		}
		size--;
		return removeNode.data;
	}

	@Override
	public String toString() {
		return "_LinkedList{" + "head=" + head + ", last=" + last + ", size=" + size + '}';
	}
}

File: golang/basic/datastruct/linked_list/linked_list.go
package linked_list

import "fmt"

type Node[T any] struct {
	Data T
	Next *Node[T]
}

type LinkedList[T any] struct {
	Head *Node[T]
	Last *Node[T]
	size int
}

// GetNode 获取指定位置的节点
func (l *LinkedList[T]) GetNode(index int) *Node[T] {
	if index > l.size {
		panic("超出链表最大索引")
	}

	node := l.Head
	for i := 0; i < index; i++ {
		node = node.Next
	}
	return node
}

// Add 在链表末尾插入元素值
func (l *LinkedList[T]) Add(e T) {
	l.AddByIndex(e, l.size)
}

// AddByIndex 在指定位置插入值
func (l *LinkedList[T]) AddByIndex(e T, index int) {
	if index > l.size {
		panic("超出链表最大索引")
	}

	newNode := &Node[T]{Data: e}
	if index == 0 {
		newNode.Next = l.Head
		l.Head = newNode
		if l.Last == nil {
			l.Last = newNode
		}
	} else if index == l.size {
		l.Last.Next = newNode
		l.Last = newNode
	} else {
		preNode := l.GetNode(index - 1)
		newNode.Next = preNode.Next
		preNode.Next = newNode
	}
	l.size++
}

// RemoveByIndex 删除指定位置的元素
func (l *LinkedList[T]) RemoveByIndex(index int) (oldNode *Node[T]) {
	preNode := l.GetNode(index - 1)
	oldNode = preNode.Next
	preNode.Next = oldNode.Next
	l.size--
	return
}

func (l *LinkedList[T]) String() string {
	var dataStr string
	node := l.Head
	for node != nil {
		dataStr += fmt.Sprintf("%s ", node.Data)
		node = node.Next
	}
	return fmt.Sprintf("data: %s, size: %d\n", dataStr, l.size)
}

func NewLinkedList[T any]() *LinkedList[T] {
	return &LinkedList[T]{}
}
